### PR TITLE
Add phpcs linting and code coverage via codecov.io

### PR DESCRIPTION
**Pull request by madmatt**
Add phpcs linting and code coverage via codecov.io. Linting supports both PHP 7.4 and 8.0.

**Changes:**
- No longer catch exceptions in GlobalAuthenticationService->verifyIdentity() - we just immediately re-threw it anyway
- Allow both $gateway and $soapOptions arguments to GlobalAuthenticationService constructor to be optional (backwards compatibility with 0.2, and make ->setApiUsername() etc actually useful)
- Add return type hints where-ever they were missing
- Add phpdocs where-ever they were missing
- Don't pass `null` into SoapVar object creation as it should be a string
- Lots of other changes to comply with linting rules (no significant code changes)
- Ensure all files are using strict_types=1

**Tests & linting:**
- Include strict slevomat/coding-standard module for additional phpcs checks
- Configure phpcs.xml.dist with specific linting rules (based on PSR-12 and slevomat/coding-standard defaults, with minor tweaks)
- Run full linting pass
- Approx ~150 changes to comply with linting rules
- Add code coverage to phpunit run
- Via GitHub Actions, upload code coverage information to codecov.io
- Include php-parallel-lint and phpcs to GitHub Actions workflow, rename from phpunit.yml to run-tests.yml
- Allow BandText to be changed during test runs and add a new test for it

**Documentation:**
- Move the full PHP code example from README.md into new docs/full-code-example.md and link through
- Add new documentation covering various module settings to README.md
- Update full code example with descriptions for key classes, and remove outdated info

**Minor:**
- Update .gitignore to ignore new .artifacts directory and re-order alphabetically

### CODE REVIEW

**Comment by madmatt** [comment]
Hey @elnurvl, if you have some time it would be great if you could test out this PR.

It adds linting support, but does make some changes. All changes _should_ be backwards compatible with version 1.0.2 that was just released, but I might have messed up somewhere, so a second pair of eyes would be awesome. Only if you have time though - otherwise I'll leave this as a PR for a while until I can get a chance to fully test it in a staging and production environment myself.

If so, I'll likely release this as version 1.1.0 given the amount of changes made to get `phpcs` happy.

File: src/Gateway/GlobalAuthenticationGateway.php
Comment (elnurvl) [inline src/Gateway/GlobalAuthenticationGateway.php]: Keeping the older parameter names would help on the code readability. Actually, I do not see any benefit of shortening names here.

File: src/Identity/Documents/InternationalPassport.php
Comment (elnurvl) [inline src/Identity/Documents/InternationalPassport.php:128]: A null check on the `$expiryDate` is necessary before these lines.

File: src/Identity/Documents/InternationalPassport.php
Comment (elnurvl) [inline src/Identity/Documents/InternationalPassport.php:180]: The same applies here as well.

File: src/Service/GlobalAuthenticationService.php
Comment (elnurvl) [inline src/Service/GlobalAuthenticationService.php]: A default `null` value should be set for the `$gateway` variable if the constructor is intended to be used without giving any parameter.

File: tests/Gateway/Request/AuthenticateSPRequestTest.php
Comment (elnurvl) [inline tests/Gateway/Request/AuthenticateSPRequestTest.php:222]: Is the `setUp` method supposed to be in the end of a test file by any guideline? I used to see it in the top.

**Review by elnurvl — CHANGES_REQUESTED** [review]
Great work! Check my comments please. Looking forward to see this being merged. But please test the changes. I just reviewed the code by reading it. Will test it and let you know if there is any other issue.

One additional request would be to atomize commits as much as you can. It would make it a lot easier to digest big changes like this.

File: tests/Gateway/Request/AuthenticateSPRequestTest.php
Comment (Firesphere) [reply tests/Gateway/Request/AuthenticateSPRequestTest.php:222]: There's no standard or guideline as far as I know of. I only know that auto-formatting with PHPStorm will push it to the bottom, from the logic of "A human wants to see the tests themselves, set-up is not the first thing they should see", if I recall correctly.

It's a matter of preference (or lazyness, whichever you prefer :) )

File: src/Gateway/GlobalAuthenticationGateway.php
Comment (madmatt) [reply src/Gateway/GlobalAuthenticationGateway.php]: Hey - yeah I agree, this was to get the line length under 120 characters, and just so I could avoid making a multi-line constructor as I just personally don't like the readability of those.

I'll adjust this (and the other places I've changed it) to be a multi-line constructor instead :)

File: src/Identity/Documents/InternationalPassport.php
Comment (madmatt) [reply src/Identity/Documents/InternationalPassport.php:128]: Good point. Looking back at the history of the method this previously took separate values that were all required ints, I think rather than adding null checks (and then either returning `null` or throwing an `Exception`, neither of which feels like a good option for a setter method), I think we should just enforce that DateTime exists at a language level by making it non-nullable.

I've made that change instead, let me know if you have a use case where you want to pass in a null `DateTime`?

FYI this is technically a breaking change but I think given how quickly we are releasing these new versions and the likelihood that this is actually being used in such a way it's unlikely, I won't release a v2.0.0 just for this fix.

File: src/Identity/Documents/InternationalPassport.php
Comment (madmatt) [reply src/Identity/Documents/InternationalPassport.php:180]: Thanks - have done the same as above.

File: tests/Gateway/Request/AuthenticateSPRequestTest.php
Comment (madmatt) [reply tests/Gateway/Request/AuthenticateSPRequestTest.php:222]: The only reason I moved it was because the `slevomat/coding-standard` comes with the ﻿[SlevomatCodingStandard.Classes.ClassStructure](https://github.com/slevomat/coding-standard#slevomatcodingstandardclassesclassstructure-) sniff which by default requires `public` methods to be above `protected` methods, and `protected` above `private`.

(Note: This is also why some other methods were moved around on other classes)

I like having `setUp` at the top too as it describes what happens before every test, but I didn't want to disable that sniff as it's generally a good idea to separate out public/protected/private methods, and you can't disable that sniff for this specific use case unfortunately.

My feeling is that it's better to live with this minor issue than remove the sniff, especially given `setUp` isn't actually doing anything ground-breaking or crazy here.

File: src/Service/GlobalAuthenticationService.php
Comment (madmatt) [reply src/Service/GlobalAuthenticationService.php]: Good spotting, my mistake for missing that. Ta!

**Comment by madmatt** [comment]
Hey @elnurvl, thanks for your quick and thorough feedback. I would usually squash all commits when opening pull requests, but fair point on this one - I should have kept it at one commit per item (e.g. for each linting change I made) to make it easier to review, sorry about that!

**Comment by madmatt** [comment]
I'm happy to leave this open until I can test it myself, but if you end up wanting to test it out and it works for you I'm also happy to merge it and tag a 1.1.0 release so you can use that sooner if you'd prefer.

File: tests/Gateway/Request/AuthenticateSPRequestTest.php
Comment (elnurvl) [reply tests/Gateway/Request/AuthenticateSPRequestTest.php:222]: Fair points.

File: src/Identity/Documents/InternationalPassport.php
Comment (elnurvl) [reply src/Identity/Documents/InternationalPassport.php:128]: I made DateTime fields to be nullable for the sake of consistency, so you can set the fields with a same logic confidently. I made the decision also based on a fact that date fields are nullable as per the WSDL.

Comment (madmatt) [reply src/Identity/Documents/InternationalPassport.php:128]: Alright, that makes sense to me - I've configured it so that if the value is null then it resets all fields that actually get sent to ID3global to null, I think this makes the most sense (rather than e.g. returning after doing nothing or throwing an Exception) - what do you think?

Comment (elnurvl) [reply src/Identity/Documents/InternationalPassport.php:128]: Yes, you are absolutely right. I did not take into account the reset logic.

**Comment by elnurvl** [comment]
@madmatt , it makes sense to squash commits by the way, but not during the creation of the pull request, rather during the merge.

File: src/Identity/Documents/InternationalPassport.php
Comment (madmatt) [reply src/Identity/Documents/InternationalPassport.php:128]: Cool, I'll resolve this conversation now then. Will look to merge this once I can test it on at least the pilot environment (my account is currently disabled :))

**Comment by elnurvl** [comment]
@madmatt , additionally the `pull_request` trigger should be enough if you want to run all checks for every subsequent commits in a pull request.

**Comment by madmatt** [comment]
Yeah I agree - I left it as push (as well as pull_request) for now just because I figured this would take a while to merge and I wanted to prove it was working from the first push, plus I don't expect this module to get a significant amount of pushes after the next few weeks as we're testing and fixing the latest version so I'm not too worried about an over-abundance of actions running :)

But yeah - will look at changing that in the future, good reminder!

**Comment by madmatt** [comment]
I have tested this work now and it all works as expected, so I'm merging this into the base branch and releasing a new version of the module :-)

**Pull request merged**
